mypy/plugins/attrs.py

```diff
@@ -709,6 +711,22 @@ def _add_init(ctx: 'mypy.plugin.ClassDefContext', attributes: List[Attribute],
     adder.add_method('__init__', args, NoneType())
 
 
+def _add_attrs_magic_attribute(ctx: 'mypy.plugin.ClassDefContext') -> None:
+    attr_name = '__attrs_attrs__'
+    any_type = AnyType(TypeOfAny.explicit)
+    attribute_type = ctx.api.named_type_or_none('attr.Attribute', [any_type]) or any_type
+    var = Var(name=attr_name, type=ctx.api.named_type('__builtins__.tuple', [
```

Comment:
Looks like this generates `Tuple[attr.Attribute[Any], ...]`. Since you know what the fields are, shouldn't it be possible to generate the precise type for each attribute?

Reply:
Maybe it is possible, but I don’t think that it is necessary to solve the original problem. It can be implemented later in another MR without breaking changes.

Reply:
Does it generate `Tuple[attr.Attribute[Any], ...]` or `Tuple[attr.Attribute[Any]]`? I don't see how the difference is specified in the mypy API, though there obviously is a difference and mypy does handle ellipses. 

Anyway, I believe the precise attribute types could be generate by changing the following lines:
```
308: _add_attrs_magic_attribute(ctx, [node.type for attr in attributes])
714: def _add_attrs_magic_attribute(ctx: 'mypy.plugin.ClassDefContext', raw_attr_types: 'List[mypy.types.Type]') -> None:
718: attribute_types = [attr_type if attr_type is not None else any_type for attr_type in raw_attr_types]
719: var = Var(name=attr_name, type=ctx.api.named_type('__builtins__.tuple', attribute_types))
```

Sort of sloppy formatting but hopefully gets the picture across. The checks for `ctx.api.final_iteration` and `ctx.api.defer` calls are completed by the time the call to `_add_attrs_magic_attribute`, so it looks safe to check for `None` and substitute `any_type` at that point.

Reply:
@EdwardLarson I changed the method to generate the exact types for the attributes. Thank you for your help.